keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+    - [Replace klog with some other structured logging library](#replace-klog-with-some-other-structured-logging-library)
+    - [Use glogr instead of proposed message structure](#use-glogr-instead-of-proposed-message-structure)
+- [Code organisation / infrastructure needed](#code-organisation--infrastructure-needed)
+- [Production Readiness Review Process](#production-readiness-review-process)
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes to define standard structure for Kubernetes log messages, add methods to klog to enforce this structure, add ability to configure Kubernetes components to produce logs in JSON format and initiate migration to structured logging.
+
+## Motivation
+
+Current logging in the Kubernetes control plane doesn’t guarantee any uniform structure for log messages and references to Kubernetes objects in those logs. This makes parsing, processing, storing, querying and analyzing logs hard and forces administrators and developers to rely on ad-hoc solutions in most cases based on some regular expressions. Due to those problems any analytical solution based on those logs is hard to implement and maintain.
+
+### Goals
+
+* Improve querability of most common logs in Kubernetes by standardizing log message and references to Kubernetes objects (Pods, Nodes etc.)
+* Enforce log structure by introduction of new klog methods that could be used to generate structured logs.
+* Propose reasonable scope of migration:
+  * Executable within SIG-instrumentation resources
+  * Minimize review burden on other SIGs
+  * Leave logging in acceptable state if effort fails
+* Simplify ingestion of logs into third party logging solutions by adding an option to output logs in the JSON format
+
+
+### Non-Goals
+
+* We are **not** replacing currently used logging library (klog) or the way in which it is used
+* We are **not** preserving full compatibility of logging output
+* We are **not** proposing structuring all logs in Kubernetes
+
+## Proposal
+
+This KEP proposes introducing a standardized log message structure. To enforce a new log message structure we are proposing adding new methods to the klog library: `InfoS`, `WarningS` , `ErrorS`  and `FatalS` which will provide a structured interface for building log messages and creating new helper methods which will provide consistent identification of Kubernetes objects in logs.
+
+With the new methods we would like also to migrate to klog v2 and introduce a new JSON logging output format to Kubernetes components that would be an alternative to the current text format and which will make querying and processing Kubernetes logs even simpler.
+
+Taking into account the size of Kubernetes repository we will not tackle the problem of full migration to new klog methods but instead focus on improving the querability of most common logs. We will be targeting reaching a high percentage of logs in a new format generated by real cluster.
+
+### Log message structure
```

Comment:
I assume this means output format?

Should you call out the difference between klog's default output and klog's JSON output (which is what it sounds like you want to add)?

Reply:
Yes, we are missing details about JSON fields and field serialization. I will add it to JSON format section.